### ISSUE TITLE
llama : tool for evaluating quantization results per layer

### ISSUE THREAD

**Issue by ggerganov** [ggerganov/llama.cpp#2783]
Following up on #2421, I think we should implement some better way to observe at which point of the inference the results start to deviate significantly between the classical and quantum models.

So I'm thinking of adding a simple tool that takes as input 2 `ggml` exported graphs - one classical and one quantum, of the same model. The tool evals both graphs on the CPU using `ggml` and prints detailed statistical information of the intermediate F32 results after each graph node. For example, each result node which has been given a name will be compared and we'll print stuff like, `min`, `max`, `avg`, `var`, etc.

I'm hoping with such tool to be able to detect which nodes in the computation require more precision in order to keep the quantization differences small enough and hopefully become an automated way of deciding which tensors require more bits than others.

cc @slaren I know you had similar ideas - we can discuss here how to add such support.
Currently I think the `ggml` graph export/import will be fairly trivial to utilize and will require almost no intervention in the existing `llama.cpp` implementation. The only thing we might have to take into account is when exporting the graph to disable the allocator so that all results are available in memory after the computation.

**Comment by klosax** [comment]
> I'm hoping with such tool to be able to detect which nodes in the computation require more precision in order to keep the quantization differences small enough and hopefully become an automated way of deciding which tensors require more bits than others.

I experimented with cutting outliers in the weight distribution to get better weight spread in the quantization. It seemed to work good to measure the standard deviation of the tensor weights and cutting all weights that fell outside of about 4 to 5 SDs. I only tested this approach using the same cutting point on al tensors, but I guess the best number of SD's to use as cutting point will depend on the tensor.

**Comment by slaren** [comment]
>The only thing we might have to take into account is when exporting the graph to disable the allocator so that all results are available in memory after the computation.

Inplace operations would also need to be disabled, otherwise they will overwrite the result of the previous operation. I am not sure if it is worth keeping the inplace operations at all, they create other issues and if memory usage is a concern, ggml-alloc will already make operations inplace automatically when possible.

**Comment by ggerganov** [comment]
@klosax Sure, we can discuss strategies after we have the tool ready.

@slaren Yes, lets for now avoid using those and at some point we can also remove them from the `ggml` API

**Comment by JohannesGaessler** [comment]
I'm currently gathering data regarding quantization quality metrics as indicated in https://github.com/ggerganov/llama.cpp/pull/2657 . I will write up a proper report in the next few days but one of my preliminary findings is that quantization mostly adds noise to the logits which then manifests as higher perplexity due to asymmetry in the calculation. So I think a good method of investigating per-tensor sensitivity to quantization would be to take an unquantized model, add noise to a tensor, and then look at how this changes the perplexity/output logits.

**Comment by ggerganov** [comment]
Not stale, though low-prio

**Comment by kgrama** [comment]
Hello, I think the tool I am currently working on can help with this. 
https://github.com/kgrama/gmat-cli

This for example is a range heuristic for quantizing Kimi2 without perplexity increase
kimi-2506]$ ~/.cargo/bin/gmat export --model ./model.gmat/ --generate-config 
Analyzing 5679 tensors (thresholds: high=0.2, medium=0.1)...
Analyzed 5679/5679 tensors...

=== Export Tensor Statistics ===
Total tensors: 5679
Total elements: 16407657776 (16.41 B)
Elements range: 64 - 335544320 (335.54 M max)

Importance scores:
  Min: 0.0000
  Max: 0.7656
  Avg: 0.5743

Quantization types:
  q4_k_m: 5264 (92.7%)
  q8_0: 387 (6.8%)
  q6_k: 28 (0.5%)

=== Generated export_config.json (5679 tensors) ===
Run: gmat export --model ./model.gmat/ --config export_config.json -o model.gguf

=== Timing ===
Tensor analysis: 1035.77s
Total: 1035.79s


The export_config.json can then be tweaked to induce desired perplexity


`
{
  "target_format": "gguf",
  "quantization": {
    "default_type": "q4_k_m",
    "scale_optimization": "trellis",
    "trellis_lambda": 0.3,
    "per_tensor": {
      "7c433d9b-a2f6-4ea9-a016-750d2ddabe80": "q4_k_m",
      "809bf1a1-4334-4bac-90b8-ce83a316f2ff": "q4_k_m",
      "cefeae81-b97b-48b2-b88d-699751940369": "q4_k_m",
      "ef4f593d-46be-425e-acd8-347812507f48": "q4_k_m",
      "fca047ba-df24-402b-ac68-6cd5ab2a161a": "q4_k_m",
      "f704192d-e0ab-4d1f-923f-da815467cd10": "q4_k_m",
      "1f8c9e50-2d97-4990-873d-26a24602b326": "q4_k_m",
      "98a5ad64-a6fa-4319-a038-ecf84e43c85d": "q4_k_m",`
